contracts/token/ERC20/extensions/draft-ERC4646.sol

```diff
@@ -0,0 +1,127 @@
+        return _sharesToAssets(shares);
+    }
+
+    function mint(uint256 shares, address receiver) public virtual override returns (uint256) {
+        uint256 assets = _sharesToAssets(shares);
+        SafeERC20.safeTransferFrom(_asset, _msgSender(), address(this), assets);
+        _mint(receiver, shares);
+        return assets;
+    }
+
+    function maxWithdraw(address caller) public view virtual override returns (uint256) {
+        return assetsOf(caller);
+    }
+
+    function previewWithdraw(uint256 assets) public view virtual override returns (uint256) {
+        return _assetsToShares(assets);
+    }
+
+    function withdraw(uint256 assets, address receiver, address owner) public virtual override returns (uint256) {
+        address sender = _msgSender();
+        uint256 shares = previewWithdraw(assets);
+
+        if (sender != owner) {
+            uint256 currentAllowance = allowance(owner, sender);
+            if (currentAllowance != type(uint256).max) {
+                require(currentAllowance >= shares, "ERC20: transfer amount exceeds allowance");
+                unchecked {
+                    _approve(owner, sender, currentAllowance - shares);
+                }
+            }
+        }
+
+        _burn(owner, shares);
+        SafeERC20.safeTransfer(_asset, receiver, assets);
+        return shares;
+    }
+
+    function maxRedeem(address caller) public view virtual override returns (uint256) {
+        return balanceOf(caller);
+    }
+
+    function previewRedeem(uint256 shares) public view virtual override returns (uint256) {
+        return _sharesToAssets(shares);
+    }
+
+    function redeem(uint256 shares, address receiver, address owner) public virtual override returns (uint256) {
+        address sender = _msgSender();
+        uint256 assets = previewRedeem(shares);
+
+        if (sender != owner) {
+            uint256 currentAllowance = allowance(owner, sender);
+            if (currentAllowance != type(uint256).max) {
+                require(currentAllowance >= shares, "ERC20: transfer amount exceeds allowance");
+                unchecked {
+                    _approve(owner, sender, currentAllowance - shares);
+                }
+            }
+        }
+
+        _burn(owner, shares);
+        SafeERC20.safeTransfer(_asset, receiver, assets);
+        return assets;
+    }
+
+    function _sharesToAssets(uint256 shares) internal view virtual returns (uint256) {
+        return totalAssets() == 0 && totalSupply() == 0
+            ? shares * (10 ** _asset.decimals()) / (10 ** decimals())
+            : shares * totalAssets() / totalSupply();
+    }
+
+    function _assetsToShares(uint256 assets) internal view virtual returns (uint256) {
+        return totalAssets() == 0 && totalSupply() == 0
+            ? assets * (10 ** decimals()) / (10 ** _asset.decimals())
+            : assets * totalSupply() / totalAssets();
```

Comment:
The contract feels almost too simple.

Can we imagine interesting customizations people will want to make?

Ideally there would be a simple way to override these conversions. What would you think of an `_exchangeRate()` function that returns a fraction expressed in as a pair `(uint256 numerator, uint256 denominator)`? The default would be `(totalSupply(), totalAssets())` (or the inverse... unclear) or something like that, properly handling 0.

Reply:
On the topic of interesting customizations, I added and am currently testing support for the native currency via additional deposit/withdrawl functions.  This assumes/requires that _asset be an implementation of WETH and uses it's deposit/withdrawl functions to automatically wrap/unwrap for the user so that _asset can still be ERC20.

Also am planning to add some unstaking duration so there is a configurable lockup period.